enhancements/builds/entitlements.md

```diff
@@ -0,0 +1,295 @@
+---
+title: openshift-builds-improve-use-of-rhel-entitlements
+
+authors:
+  - "@gabemontero"
+  - "@bparees" (provided a lot of input in tracking Jira)
+
+reviewers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@who else?"
+  
+approvers:
+  - "@bparees"
+  - "@adambkaplan"
+  - "@who else?"
+  
+creation-date: 2020-02-18
+
+last-updated: yyyy-mm-dd
+
+<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
+status: provisional
+
+see-also:
+  - "/enhancements/builds/volume-secrets.md"  
+  
+---
+
+# A More Automatic Experience For Use of Entitled RHEL Content In OpenShift V4 Builds
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+ > 1. CRI-O either already has or will have an option to specify host directories that should always be mounted 
+>into containers, akin to what the Red Hat Docker Daemon used to provide with OpenShift V3.  If / when that feature
+>is available could have bearing on the implementation options discussed below, or even when to stage in this 
+>enhancement.  Presumably the MCO would need updating to leverage this CRI-O feature.
```

Comment:
Who or what is going to provide the rhsm.conf, that is missing in RHCOS? The subscription-manager certs are installed per default. When CRI-O mounts the creds into the container the subscription-manager is looking into rhsm-host for the rhsm.conf which is missing. the current solution is to provide the rhsm.conf manually with the pem files.

Reply:
Some customers are uncomfortable to use MCO to "just" configure entitlements, they see it more like a configuration change in the cluster e.g "oc edit entitlements" "enable: true".

Reply:
The admin would provide the content to MCO.  

It is not the most natural use of MCO, but basically MCO is a way to manage the filesystem of nodes.  The fact that in this case the files are nothing RHCOS or OCP actually care about doesn't matter.

The advantage to this approach is it's very seemless to the end user, most similar to how it worked in 3.x from a user's perspective.

(and if you want the content injected into all pods, not just build pods, this also accomplishes that)

Reply:
how will this approach scale across 1k clusters?

what would we recommend our multi-cluster management experience around enabling this capability to be?  it seems easier to have an entitlements policy flag that is synchronized down to each cluster rather than have to synchronize machine configurations down to each cluster.

it should be a simple step in a future multi-cluster experience to enable an "entitled builds" policy and have that cascaded to each member cluster.

Reply:
if we chose this direction (and i think we are leaning away from it, but i wanted it listed for completeness since it is most similar to the 3.x mechanism) i would expect it could live as a patchset that could be applied to the machineconfigs of any cluster.  Maybe that's hopelessly naive as i haven't looked into the precise format of the machineconfig and the content we'd have to add to it.

Reply:
OK, I've got a change sitting on my laptop, that I've posted below, that attempts to
- correct per @bparees 's comments the fact that we don't have to "change" MCO, but rather use it if we go down that path
- list @derekwaynecarr 's and @zvonkok 's concerns
- attempt to clarify possibilities around scope of this proposal vs. possible additional proposals, and that orders of precedence need to be agreed upon if both broad and point solutions coming various stages:

If you all want to hack at it before I push the commit, while I'm looking at the other comments, here is the new text around this open questions:

```
CRI-O either already has or will have an option to specify host directories that should always be mounted 
>into containers, akin to what the Red Hat Docker Daemon used to provide with OpenShift V3.  If / when that feature
>is available could have bearing on the implementation options discussed below, or even when to stage in this 
>enhancement.  With current capabilities, the cluster admin would use the MCO post install to effectively update the
>filesystem of each node with the necessary subscription related files in a well known location that CRI-O would look
>for.  There is concern though using this through MCO from a couple of perspectives.  One perspective of concern is 
>multi-cluster management, and scaling out to many clusters.  As such, enabling entitlement, either for builds only or 
>for pods in general, could be policy bits around the multi-cluster set up experience.  Related to that, but also 
>relevant in a single cluster scenario, is around customers having a specific config resource around entitlements, 
>and engaging through that instead of the more generic MCO.  Ultimately, with respect to this proposal, which is currently
>for builds only, there is first a scope question around devising a broader architecture around entitlements, across 
>a broader range of components.  That is either a separate proposal, or a broadening of the scope of this proposal
>from its original intent.  Second, if we manage this with separate proposals, do we gate this one or make it 
>dependent on the broader proposal, or move forward with this one, and simply get agreement on what the order of 
>precedence would be when both a point solution around builds and broader solution around accessing entitlements 
>from any pod.
```

Reply:
FYI on the multi cluster front I had a quick chat with @csrwng and he pointed me to [hive sync sets](https://github.com/openshift/hive/blob/master/docs/syncset.md) as a possible pathway to @derekwaynecarr 's ask around a simple enable entitled builds policy and have it cascade to each cluster thingie

Going to study hive a bit, get about 15 to 20 minutes of voice to voice time with @csrwng and then see if I can translate that to an update here.

Reply:
@bparees @zvonkok FYI ^^

Reply:
OK @csrwng and I talked and he quickly realized that @derekwaynecarr when he says "policy" is referring to the Multi Cluster Manager stuff from the group of IBMers who just joined Red Hat (they are actually here for orientation today, impending snow be damned ;-) )

Anyway while MCM is using hive for some things, it is not using hive SyncSets for the type of enabling/injecting we are talking about here.

Cesar pointed me to some more reading at https://www.ibm.com/cloud/architecture/architectures/multicloudManagementArchitecture

I'll see what I can gleam from that and how solid of a connection I can write down here in this proposal for connecting the dots.

@bparees @zvonkok FYI

Reply:
OK I've pushed the update around multi cluster and that we have a broader "delivery mechanism" dimension to add to the original matrix of question that @bparees gave me as a starting point.

I'll update some of the pros/cons around the MCO option on the other comment thread.

I'm going to wrap up this iteration with my latest push.